### PR TITLE
[RFC][Resource] Make Resource actions easier to overwrite

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Related tickets |
| License         | MIT

As we know, ``ResourceController`` gives us a lot, starting from duplication reduction, on sense of security ending 😄  However, it's pretty hard to use its methods with some extra data or logic - it usually ends up with copying whole action. It's ridiculous, especially when we just want to add some extra data to template (which was my use case in https://github.com/Sylius/Sylius/pull/6161 - I ended up with overwriting the whole damn action).
I suggest to introduce some protected methods returning templates data, that can be overwrote instead of actions. I know it's not the most clean way (we still have some parameters passed to these functions which looks awful for method, that should just return required data), but could be a first step for making ``ResourceController`` easier to use.
I would appreciate any feedback about this, is it the path we want to follow or not. 🐫

### CODE REVIEW

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 1]
```diff
@@ -550,4 +531,87 @@ protected function findOr404(RequestConfiguration $configuration)
...
+    }
+
+    /**
+     * @param RequestConfiguration $configuration
+     * @param ResourceInterface $resource
+     * @param FormInterface $form
+     *
+     * @return array
+     */
+    protected function getUpdateTemplateData(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: We should leave it in one line.

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 2]
```diff
@@ -550,4 +531,87 @@ protected function findOr404(RequestConfiguration $configuration)
...
+    }
+
+    /**
+     * @param RequestConfiguration $configuration
+     * @param ResourceInterface $newResource
+     * @param FormInterface $form
+     *
+     * @return array
+     */
+    protected function getCreateTemplateData(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Is it really needed to different create and update template data provider?

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 3]
```diff
@@ -550,4 +531,87 @@ protected function findOr404(RequestConfiguration $configuration)
...
+            'configuration' => $configuration,
+            'metadata' => $this->metadata,
+            'resource' => $resource,
+            $this->metadata->getName() => $resource,
+        ];
+    }
+
+    /**
+     * @param RequestConfiguration $configuration
+     * @param mixed $resources
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Isn't it `ResourceInterface`?

Comment (Zales0123) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: No :( As I noticed, there could be passed array or ``ResourceGridView`` and God knows what else... 😢

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php (hunk 2 above)
Comment (Arminek) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: This method name is so confusing.

Comment (Zales0123) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Tell me more how it should be named, I have no idea 😄

Comment (lchrusciel) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: IMHO we should leave just `provideFormTemplateData`.

Comment (Zales0123) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Ok, so there are two approaches:
1. ``provideFormTemplateData`` for create and update action
2. ``provideCreate\UpdateFormTemplateData`` which uses ``provideFormTemplateData`` for templates data separation

👍  for 1, 🎉  for 2, ready, set, go! 😄

**Comment by pjedrzejewski** [comment]
From my experience, 99% cases you can obtain the extra data in a much cleaner way. (partial that is cachable separately / twig)

And if you really need custom data from the controller, then resource is not the way to go and much cleaner is to simply add a super simple custom controller. (most of the time not needed though)

So I am against this change, unless we see a really annoying use-case for that. :D I commented in #6161, which is (I assume) the cause for this PR.

**Pull request closed without merging**